python/opae.admin/opae/admin/sysfs.py

```diff
@@ -0,0 +1,282 @@
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+# * Neither the name  of Intel Corporation  nor the names of its contributors
+#   may be used to  endorse or promote  products derived  from this  software
+#   without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+# IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+# LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+# CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+# SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+# INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+# CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.a
+import glob
+import os
+import re
+from contextlib import contextmanager
+from opae.utils.process import call_process, DRY_RUN
+from opae.utils.log import loggable, LOG
+
+
+PCI_ADDRESS_PATTERN = (r'(?P<pci_address>'
+                       r'(?P<segment>[\da-f]{4}):(?P<bdf>(?P<bus>[\da-f]{2}):'
+                       r'(?P<device>[\da-f]{2})\.(?P<function>\d)))')
+PCI_ADDRESS_RE = re.compile(PCI_ADDRESS_PATTERN, re.IGNORECASE)
+
+
+class sysfs_node(loggable):
+    """sysfs_node is a base class representing a sysfs object in sysfs """
+    def __init__(self, sysfs_path):
+        super(sysfs_node, self).__init__()
+        self._sysfs_path = sysfs_path
+
+    def node(self, *nodes):
+        path = os.path.join(self._sysfs_path, *nodes)
+        if not os.path.exists(path):
+            raise NameError("Could not find sysfs node: {}".format(path))
+        return sysfs_node(path)
+
+    def find(self, pattern):
+        for p in glob.glob(os.path.join(self._sysfs_path, pattern)):
+            yield sysfs_node(p)
+
+    @contextmanager
+    def _open(self, mode):
+        try:
+            with open(self._sysfs_path, mode) as fd:
+                yield fd
+        except IOError as ioerr:
+            self.log.exception(ioerr)
+            raise
+
+    @property
+    def value(self):
+        with self._open('r') as fd:
+            return fd.read().strip()
+
+    @value.setter
+    def value(self, val):
+        if DRY_RUN:
+            print('echo {} > {}'.format(val, self._sysfs_path))
+        else:
+            with self._open('w') as fd:
+                fd.write(val)
+
+
+    @property
+    def sysfs_path(self):
+        return self._sysfs_path
+
+
+class pci_node(sysfs_node):
+    PCI_BUS_SYSFS = '/sys/bus/pci/devices'
```

Comment:
Thinking of access to our fpga's device nodes (/dev/intel-fpga-port.X, /dev/intel-fpga-fme.0), what would be the means of determining this?  Would we need to define a subclass of pci_node called intel_fpga_accelerator_node, and then provide a means to retrieve the device file name?  Since this is specific to the intel fpga device, it doesn't seem right to put them into pci_node.

Reply:
The idea is that we can have a pac class that derives from sysfs_class. Using this can create instances of card-specific classes. This pac class would know about the /dev/... interface.